libbeat/common/mapstr.go

```diff
@@ -315,10 +315,46 @@ func tryToMapStr(v interface{}) (MapStr, bool) {
 	}
 }
 
-// walkMap walks the data MapStr to arrive at the value specified by the key.
+// walkMapRecursive walks the data MapStr to arrive at the value specified by the key.
 // The key is expressed in dot-notation (eg. x.y.z). When the key is found then
 // the given mapStrOperation is invoked.
-func walkMap(key string, data MapStr, op mapStrOperation) (interface{}, error) {
+func walkMapRecursive(key string, data MapStr, op mapStrOperation) (interface{}, error) {
+
+	// Splits up the key in two parts: full key and first part before the dot
+	_, exists := data[key]
+	// If leave node or key exists directly
+	if exists {
+		// Execute the mapStrOperator on the leaf object.
+		return op.Do(key, data)
+	}
+
+	keyParts := strings.SplitN(key, ".", 2)
+	// If leave node or key exists directly
+	if len(keyParts) == 1 {
+		// Execute the mapStrOperator on the leaf object.
+		return op.Do(key, data)
+	}
+
+	// Checks if first part of the key exists
+	k := keyParts[0]
+	d, keyExist := data[k]
+	if !keyExist {
+		if op.CreateMissingKeys {
+			data[k] = MapStr{}
```

Comment:
You’ll need to set d here as well.

Reply:
Hm, good catch. Problem is when setting using `d` here instead I first have to do the type conversion / check. Let me see if I can reorder code further.

Reply:
I _think_ what you had was fine. `d` is an `interface{}` so you should be able to assign any value to it. Looks like you just need this:

```
d = MapStr{}
data[k] = d
```

Reply:
done and works.